### DISCUSSION TITLE
getStaticProps (SSG) providers at build time

### DISCUSSION THREAD

**Discussion by NuroDev** [nextauthjs/next-auth#4823] [Help]
I've currently got a custom login page set up using `getStaticProps` (SSG / ISR) to call `getProviders` & pass it to the client.
Similar to if I set it up using `getServerSideProps` (SSR) I want to dynamically fetch providers, but to me I will not be changing my providers that often such that I feel it would be better either just making it a build step (SSG) or update it on a set interval (ISR).

```ts
export const getStaticProps: GetStaticProps = async () => {
	const providers = await getProviders();

	return {
		props: {
			providers,
		},
		revalidate: 1800,
	};
};
```

However, the one issue I have found is that `getProviders` requires an active endpoint to hit to retrieve the providers which for SSG means it's not possible. The only work around I've found for the time being is to set the `NEXTAUTH_URL` to the production URL at build time which is not ideal.

From here do I:
- Just switch back to using `getServerSideProps` (SSR)
- Use some fix / alternative I was not yet aware of
- Create a feature request for something like `getServerProviders`, similar to `getServerSession` which directly invokes the providers endpoint logic for server side use only?

**Comment by viperxeyes** [comment]
I know this is old , but I faced this today and I came up with two solutions:

1-I just called the function in useEffect and converted the result to array and mapped through it  

``` ts
useEffect(() => {
    const loadProviders = async () => {
      const res = await getProviders();
      const providersList = Object.keys(res).map((k) => res[k]);
      setProviders(providersList);
    };
    loadProviders();
  }, []);
```
then map through the array and  to return the buttons. but wiht this solution there will be layour shift since it is client rendered .


2-ince i know i won't change my providers much , i save the output of getProviders in a defined constant array 

```ts
const [providers, setProviders] = useState([
    {
      id: "credentials",
      name: "Credentials",
      type: "credentials",
    },
    {
      id: "google",
      name: "Google",
      type: "oauth",
    },
    {
      id: "facebook",
      name: "Facebook",
      type: "oauth",
    },
    {
      id: "discord",
      name: "Discord",
      type: "oauth",
    },
    {
      id: "twitter",
      name: "Twitter",
      type: "oauth",
    },
  ]);
```
and do the samething as solution one from here on But the only difference is my signin methog  i pass the callback options

```ts

onClick={async (e) => {
                           await signIn(provider.id, {
                            callbackUrl: router.query.callbackUrl
                              ? router.query.callbackUrl
                              : router.route,
                          
                          });
                       
                        }}
```
if there is better solution please let us know

**Reply by NuroDev** [reply]
I actually ended up solving this by re-creating the server-side logic for the NextAuth.js providers endpoint so I can use it anywhere just by passing in the providers for my NextAuth.js config / options:
```typescript
import type { Provider } from 'next-auth/providers';
import type { PublicProvider } from 'next-auth/core/routes/providers';

/**
 * Get Server Providers
 *
 * @description Alternative to `getProviders` client method from NextAuth.js designed
 * specifically for use on the server / at build time (Primarily SSG).
 *
 * @param {Array} providers - Array of auth providers from your NextAuth.js config
 */
export function getServerProviders(providers: Array<Provider>): Record<string, PublicProvider> {
	// Note: Deployment provider failover URL to add preview environment support
	const baseUrl = process.env.NEXTAUTH_URL || `https://${process.env.VERCEL_URL}`;

	return Object.fromEntries(
		providers.map(({ id, name, type }): [string, PublicProvider] => [
			id,
			{
				callbackUrl: new URL(`/api/callback/${id}`, baseUrl).href,
				id,
				name,
				signinUrl: new URL(`/api/signin/${id}`, baseUrl).href,
				type,
			},
		]),
	);
}
```

With the function above you can get providers to pass as props for SSG:
```tsx
import { getServerProviders } from "@apply/auth/providers";
import { signIn } from "next-auth/react";

import type { GetStaticProps } from "next";
import type { PublicProvider } from "next-auth/core/routes/providers";

interface LoginPageProps {
  providers: Record<string, PublicProvider>;
}

export default function LoginPage({ providers }: LoginPageProps) {
  return (
    <>
      {Object.values(providers).map((provider) => (
        <button
          key={provider.id}
          onClick={() => signIn(provider.id)}
          type="button"
        >
          Sign in with {provider.name}
        </button>
      ))}
    </>
  );
}

export const getStaticProps: GetStaticProps<LoginPageProps> = async () => {
  return {
    props: {
      providers: getServerProviders(),
    },
  };
};
```

**Comment by Theodlz** [comment]
Hi @NuroDev. I'm trying to build a webapp that uses ISR and next-auth. On the frontpage page `index.js` I wanna use a `getStaticProps` to revalidate every 60 seconds and rerender the page by calling an internal API that gets the data from a mongdb cluster. However, this is giving me errors at build time, telling me that it can't fetch `/api/auth/session`, which doesn't seem to happen to you. In the `getStaticProps` I am not even calling any session or whatever, but is looks like its looking for it as soon as I add a `getStaticProps` with an API call in it.

Do you have any idea what is causing it?

**Reply by Theodlz** [reply]
With a bit more of trial an error, it seems that if I use an API that is not mine (where I also have a NextAuth endpoint) it does not cause and error. It seems that somehow NextAuth "wraps" my whole API and triggers a call to its own endpoint as soon as I hit any other endpoint. Even those that do not have any next-auth code in them.

**Reply by NuroDev** [reply]
> It seems that somehow NextAuth "wraps" my whole API and triggers a call to its own endpoint

What way have you implemented NextAuth.js? Is it via middleware by chance? If so that would make sense why any & all requests are being routed through NextAuth.js & are requiring a valid auth session.
